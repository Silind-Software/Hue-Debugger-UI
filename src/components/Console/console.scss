@import '../../styles/main.scss';

.consoleWrapper {
  width: 100%;
  background-color: $shade3;
  position: absolute !important;
  bottom: 0;
  left: 0;
  overflow: hidden;
  transition: 0.1s;
  z-index: 300;
}

.consoleHeader {
  @include no-select;
  width: 100%;
  height: 30px;
  background-color: $base1;
  text-align: center;
  line-height: 30px;
  color: $base2;
  font-size: 16px;
  cursor: pointer;
}

.consoleHeaderShow {
  background-color: $base5;
  color: $base3;
}

.outputWrapper {
  white-space: pre;
  margin-bottom: 10;
  transition: 0.15s;
  color: $base5;

  &:hover {
    color: $base2;
  }
}

.menuLine {
  @include no-select;
  height: 35px;
  position: absolute;
}

.menuButton {
  width: 35px;
  height: 35px;
  line-height: 35px;
  text-align: center;
  color: $base5;
  font-size: 30px;
  cursor: pointer;
  @include toolTip {
    content: attr(data-title);
  }
}

.consoleContent {
  height: calc(100% - 50px);
  padding: 20px 65px;
  overflow-y: auto;
  color: $base5;
  font-size: 13; 
  font-family: monospace;
}

.consoleTable td:nth-of-type(1) {
  width: 100px;
}