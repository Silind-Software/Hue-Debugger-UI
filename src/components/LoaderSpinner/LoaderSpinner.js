import React from 'react';
import AuthButton from '../AuthButton/AuthButton';
import "./loaderSpinner.scss";

/**
 * Loader spinner
 * Shows a loader spinner animation while loading.
 * When done loading shows a message and a back-button
 * @prop {boolean} isLoading show spinner if true, shows back-action otherwise
 * @prop {string} failMessage message shown if failed
 * @prop {function} backAction function that will be called on back-action 
 */
const LoaderSpinner = props => {
  if (props.isLoading)
    return (
      <div className="loaderSpinnerWrapper">
        <div className="lds-ripple">
          <div></div>
          <div></div>
        </div>
        <div className="loaderSpinnerText">Loading Hue resources</div>
      </div>
    )
  else
    return (
      <div className="loaderSpinnerWrapper">
        <div className="loaderSpinnerText">
          An error occured:<br/>
          {props.failMessage}
        </div>
        <AuthButton 
          hint="Re-authenticate Hue"
          onClick={props.backAction}
        />
      </div>
    )
}

export default LoaderSpinner;