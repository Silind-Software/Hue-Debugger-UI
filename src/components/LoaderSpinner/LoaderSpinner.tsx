import React, { FC } from 'react';
import AuthButton from '../AuthButton/AuthButton';
import './loaderSpinner.scss';

interface IProps {
  failMessage: string;
  isLoading: boolean;
  backAction: () => void;
}

const LoaderSpinner: FC<IProps> = (props) => {
  if (props.isLoading) {
    return (
      <div className='loaderSpinnerWrapper'>
        <div className='lds-ripple'>
          <div />
          <div />
        </div>
        <div className='loaderSpinnerText'>Loading Hue resources</div>
      </div>
    );
  }

  return (
    <div className='loaderSpinnerWrapper'>
      <div className='loaderSpinnerText'>
        An error occured:
        <br />
        {props.failMessage}
      </div>
      <AuthButton hint='Re-authenticate Hue' onClick={props.backAction} />
    </div>
  );
};

export default LoaderSpinner;
