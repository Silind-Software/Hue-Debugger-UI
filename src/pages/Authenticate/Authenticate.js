import React, { Component } from "react";
import * as service from "../../services/xs-json-requester";
import AuthButton from '../../components/AuthButton/AuthButton';
import "./authenticate.scss";

class Authenticate extends Component {
  state = {
    inputIp: "",
    inputDevName: "",
    authenticateWaiter: false,
    authenticateTimeout: null,
  };

  validateInput = async () => {
    await this.setState({ inputIp: this.state.inputIp.replace(/[a-zA-Z]|:|\//g, "") });

    if (!this.state.inputIp)
      this.props.showSweetAlertWarning(
        "Fill the IP address of your HUE bridge"
      );
    else if (!this.validateIfIp(this.state.inputIp))
      this.props.showSweetAlertWarning("Doesn't seem like that IP is right");
    else if (!this.state.inputDevName)
      this.props.showSweetAlertWarning("Choose a cool developer name");
    else if (!this.validateDevName(this.state.inputDevName))
      this.props.showSweetAlertWarning("Developer name is not accepted");
    else this.beginCountdown();
  };

  startAuthenticateProbing = () => {
    const probe = setInterval(() => {
      if (this.state.authenticateWaiter)
        this.authorizeNewUser();
      else
        clearInterval(probe);
    }, 1500);
  }

  authorizeNewUser = () => {
    service
      .postJSON("http://" + this.state.inputIp + "/api", {
        devicetype: "hue_debugger_ui#" + this.state.inputDevName
      })
      .then(res => (
        Object.keys(res[0]).forEach(key => {
          if (key === "success") {
            clearTimeout(this.state.authenticateTimeout);
            this.props.setAuthentication(
              this.state.inputIp,
              res[0]["success"]["username"]
            )
          }
          else 
            return;
        }
      )))
      .catch(e => console.log(e.message))
  };

  renderInputFields = () => (
    <div className="authContainer">
      <div className="authHeadline">Enter bridge IP</div>
      <input
        value={this.state.inputIp}
        onChange={this.onInputChange("inputIp")}
        type="text"
        className="authInput"
        placeholder="e.g. 192.168.0.100"
      />

      <div className="authHeadline">Developer name</div>
      <input
        value={this.state.inputDevName}
        onChange={this.onInputChange("inputDevName")}
        type="text"
        className="authInput"
        placeholder="Choose a developer name"
      />

      <AuthButton 
        hint="Begin"
        onClick={this.validateInput}
      />
      
    </div>
  );

  renderCountdown = () => (
    <div className="authContainer">
      <div className="countdownLine">
        <div className="countdownProcess" />
      </div>

      <div className="bridgePushRow">
        <div className="bridgePushIcon" />
        <div className="bridgePushText">
          Go and press the Hue Bridge button to authenticate!
        </div>
      </div>
    </div>
  );

  validateIfIp = sequence => {
    let isIp = true;
    sequence.split(".").forEach(block => {
      if (isNaN(block) || block < 0 || block > 255) isIp = false;
    });
    return isIp;
  };

  validateDevName = name => {
    return /^[a-zA-ZæøåÆØÅ]+$/gm.test(name);
  };

  onInputChange = input => e => this.setState({ [input]: e.target.value });

  beginCountdown = async () => {
    await this.setState({ authenticateWaiter: true });
    await this.setState({ authenticateTimeout: setTimeout(() => this.endCountdown(), 15500) });
    this.startAuthenticateProbing();
  };

  endCountdown = async () => {
    await this.setState({ authenticateWaiter: false });
    this.props.showSweetAlertWarning("Hue failed to authenticate!\nDid you press the button on the bridge?");
  }

  render() {
    if (this.state.authenticateWaiter)
      return this.renderCountdown();
    else
      return this.renderInputFields();
  }
}

export default Authenticate;
