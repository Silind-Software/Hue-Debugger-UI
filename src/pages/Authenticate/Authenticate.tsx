import React, { FC, useState, FormEvent } from 'react';
import * as service from '../../services/xs-json-requester';
import AuthButton from '../../components/AuthButton/AuthButton';
import './authenticate.scss';

interface IProps {
  setAuthentication: (ip: string, devName: string) => void;
  showSweetAlertWarning: (text: string) => void;
}

const Authenticate: FC<IProps> = (props) => {
  const [inputIp, setInputIp] = useState<string>('');
  const [inputDevName, setInputDevName] = useState<string>('');
  const [authenticateWaiter, setAuthenticateWaiter] = useState<boolean>(false);
  const [authenticateTimeout, setAuthenticateTimeout] = useState<NodeJS.Timeout | null>(null);

  const onInputIpChange = (changer: React.Dispatch<React.SetStateAction<string>>) => {
    return (e: FormEvent<HTMLInputElement>) => {
      changer(e.currentTarget.value);
    };
  };

  const validateInput = () => {
    const formattedInput = inputIp.replace(/[a-zA-Z]|:|\//g, '');

    if (!formattedInput) {
      props.showSweetAlertWarning('Fill the IP address of your HUE bridge');
      return;
    }

    if (!validateIfIp(formattedInput)) {
      props.showSweetAlertWarning('That IP is not right');
      return;
    }

    if (!inputDevName) {
      props.showSweetAlertWarning('Choose a cool developer name');
      return;
    }

    if (!validateDevName(inputDevName)) {
      props.showSweetAlertWarning('Developer name is not accepted');
      return;
    }

    beginCountdown();
    setInputIp(formattedInput);
  };

  const startAuthenticateProbing = () => {
    const probe = setInterval(() => {
      if (authenticateWaiter) {
        authorizeNewUser();
      } else {
        clearInterval(probe);
      }
    }, 1500);
  };

  const authorizeNewUser = async () => {
    try {
      const request = await service.postJSON(`http://${inputIp}/api`, {
        devicetype: `hue_debugger_ui#${inputDevName}`,
      });

      const result = await request.json();

      Object.keys(result[0]).forEach((key) => {
        if (key === 'success') {
          if (authenticateTimeout) {
            clearTimeout(authenticateTimeout);
          }
          props.setAuthentication(inputIp, result[0].success.username);
        }
      });
    } catch (error) {
      console.log(error.message);
    }
  };

  const renderInputFields = () => (
    <div className='authContainer'>
      <div className='authHeadline'>Enter bridge IP</div>
      <input
        value={inputIp}
        onChange={onInputIpChange(setInputIp)}
        type='text'
        className='authInput'
        placeholder='e.g. 192.168.0.100'
      />

      <div className='authHeadline'>Developer name</div>
      <input
        value={inputDevName}
        onChange={onInputIpChange(setInputDevName)}
        type='text'
        className='authInput'
        placeholder='Choose a developer name'
      />

      <AuthButton hint='Begin' onClick={validateInput} />
    </div>
  );

  const renderCountdown = () => (
    <div className='authContainer'>
      <div className='countdownLine'>
        <div className='countdownProcess' />
      </div>

      <div className='bridgePushRow'>
        <div className='bridgePushIcon' />
        <div className='bridgePushText'>Go and press the Hue Bridge button to authenticate!</div>
      </div>
    </div>
  );

  const validateIfIp = (sequence: string) => {
    let isIp = true;
    sequence.split('.').forEach((block: string) => {
      const parsedBlock = Number(block);
      if (isNaN(parsedBlock) || parsedBlock < 0 || parsedBlock > 255) {
        isIp = false;
      }
    });

    return isIp;
  };

  const validateDevName = (name: string) => {
    return /^[a-zA-ZæøåÆØÅ]+$/gm.test(name);
  };

  const beginCountdown = async () => {
    setAuthenticateWaiter(true);
    setAuthenticateTimeout(setTimeout(endCountdown, 15500));
    startAuthenticateProbing();
  };

  const endCountdown = async () => {
    setAuthenticateWaiter(false);
    props.showSweetAlertWarning(
      'Hue failed to authenticate!\nDid you press the button on the bridge?',
    );
  };

  if (authenticateWaiter) {
    return renderCountdown();
  }

  return renderInputFields();
};

export default Authenticate;
