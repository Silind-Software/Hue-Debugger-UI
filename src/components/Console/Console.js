import React, { Component } from "react";
import * as moment from "moment";
import "moment/locale/pt-br";
import "./console.scss";

class Console extends Component {
  state = {
    showFormatted: true
  };

  consoleContainerRep = React.createRef();

  componentDidUpdate(prevProps) {
    if (prevProps.consoleOutput !== this.props.consoleOutput)
        this.scrollToBottom();
  }

  scrollToBottom = () => {
    this.consoleContainerRep.current.scrollTo(0, this.consoleContainerRep.current.scrollHeight);
  }

  toggleFormatting = () =>
    this.setState({ showFormatted: !this.state.showFormatted });

  consoleOutput = () =>
    this.props.consoleOutput.map((o, index) => (
      <ConsoleOutput
        key={index}
        output={o}
        formatted={this.state.showFormatted}
      />
    ));

  render() {
    const formattedIcon = this.state.showFormatted ? "notes" : "sort";
    return (
      <div
        className={
          this.props.show
            ? "consoleWrapper consoleWrapperShow"
            : "consoleWrapper"
        }
      >
        <div
          className={
            this.props.show
              ? "consoleHeader consoleHeaderShow"
              : "consoleHeader"
          }
          onClick={this.props.toggleConsole}
        >
          Console
        </div>
        <div className="menuLine">
          <div 
            className="menuButton" 
            onClick={this.toggleFormatting}
            data-title={this.state.showFormatted ? "No formatting" : "Formatting"}  
          >
            <i className="material-icons">{formattedIcon}</i>
          </div>
        </div>
        <div
          ref={this.consoleContainerRep} 
          className="consoleContent"
        >
          {this.consoleOutput()}
        </div>
      </div>
    );
  }
}

class ConsoleOutput extends Component {
  state = {
    now: moment().format("h:mm:ss a")
  };

  formattedConsoleOutput = () => {
    const rawConsoleOutput = () => (
      <tbody>
        {this.formattedConsoleRow(this.state.now, "")}
        {this.formattedConsoleRow(
          "",
          JSON.stringify(this.props.output, null, 2)
        )}
      </tbody>
    );

    const outputContent = () => {
      if (this.props.output[0] === undefined)
        return (
          <tbody>
            {this.formattedConsoleRow(
              this.state.now,
              "error: undefined output"
            )}
          </tbody>
        );

      const logType = Object.keys(this.props.output[0])[0];
      if (logType === "success")
        return (
          <tbody>
            {this.formattedConsoleRow(this.state.now, "success:")}
            {this.formattedConsoleRow(
              "",
              JSON.stringify(this.props.output[0][logType], null, 2)
            )}
          </tbody>
        );
      else
        return (
          <tbody>
            {this.formattedConsoleRow(
              this.state.now,
              "error: type " + this.props.output[0][logType].type
            )}
            {this.formattedConsoleRow(
              "",
              "address: " + this.props.output[0][logType].address
            )}
            {this.formattedConsoleRow(
              "",
              "description: " + this.props.output[0][logType].description
            )}
          </tbody>
        );
    };

    if (this.props.formatted)
      return <table className="consoleTable">{outputContent()}</table>;
    else return <table className="consoleTable">{rawConsoleOutput()}</table>;
  };

  formattedConsoleRow = (firstCell, secondCell) => (
    <tr>
      <td>{firstCell}</td>
      <td>{secondCell}</td>
    </tr>
  );

  render() {
    return <div className="outputWrapper">{this.formattedConsoleOutput()}</div>;
  }
}

export default Console;
