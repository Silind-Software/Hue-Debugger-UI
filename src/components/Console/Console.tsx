import React, { FC, useState, useEffect } from 'react';
import moment from 'moment';
import { Resizable } from 're-resizable';
import 'moment/locale/pt-br';
import './console.scss';

interface IConsoleProps {
  consoleOutput: any[];
  show: boolean;
  toggleConsole: () => void;
}

const Console: FC<IConsoleProps> = (props) => {
  const [showFormatted, setShowFormatted] = useState<boolean>(true);
  const [currentHeight, setCurrentHeight] = useState<number>(300);
  const consoleContainerRep = React.useRef<HTMLDivElement>(null);

  useEffect(() => {
    scrollToBottom();
  }, [props.consoleOutput]);

  const scrollToBottom = () => {
    consoleContainerRep.current?.scrollTo(0, consoleContainerRep.current.scrollHeight);
  };

  const toggleFormatting = () => {
    setShowFormatted((prevShowFormatted: boolean) => !prevShowFormatted);
  };

  const consoleOutput = () => {
    return props.consoleOutput.map((o, index) => (
      <ConsoleOutput key={index} output={o} formatted={showFormatted} />
    ));
  };

  const formattedIcon = showFormatted ? 'notes' : 'sort';
  const headerClasses = props.show ? 'consoleHeader consoleHeaderShow' : 'consoleHeader';
  const dataTitle = showFormatted ? 'No formatting' : 'Formatting';

  const handleChangeHeight = (e: any, dir: any, ref: any, d: { width: number; height: number }) => {
    setCurrentHeight((prevHeight) => prevHeight + d.height);
  };

  return (
    <Resizable
      className='consoleWrapper'
      size={{ width: '100%', height: props.show ? currentHeight : 30 }}
      onResizeStop={handleChangeHeight}
      enable={{ top: props.show }}
      maxHeight={760}
      minHeight={30}
    >
      <div>
        <div className={headerClasses} onClick={props.toggleConsole}>
          Console
        </div>
      </div>
      <div className='menuLine'>
        <div className='menuButton' onClick={toggleFormatting} data-title={dataTitle}>
          <i className='material-icons'>{formattedIcon}</i>
        </div>
      </div>
      <div ref={consoleContainerRep} className='consoleContent'>
        {consoleOutput()}
      </div>
    </Resizable>
  );
};

interface IConsoleOutputProps {
  output: any;
  formatted: boolean;
}

const ConsoleOutput: FC<IConsoleOutputProps> = (props) => {
  const now = moment().format('h:mm:ss a');

  const formattedConsoleOutput = () => {
    const rawConsoleOutput = () => (
      <tbody>
        {formattedConsoleRow(now, '')}
        {formattedConsoleRow('', JSON.stringify(props.output, null, 2))}
      </tbody>
    );

    const outputContent = () => {
      if (props.output[0] === undefined) {
        return <tbody>{formattedConsoleRow(now, 'error: undefined output')}</tbody>;
      }

      const logType = Object.keys(props.output[0])[0];

      if (logType === 'success') {
        return (
          <tbody>
            {formattedConsoleRow(now, 'success:')}
            {formattedConsoleRow('', JSON.stringify(props.output[0][logType], null, 2))}
          </tbody>
        );
      }

      return (
        <tbody>
          {formattedConsoleRow(now, `error: type  ${props.output[0][logType].type}`)}
          {formattedConsoleRow('', `address:  ${props.output[0][logType].address}`)}
          {formattedConsoleRow('', `description:  ${props.output[0][logType].description}`)}
        </tbody>
      );
    };

    if (props.formatted) {
      return <table className='consoleTable'>{outputContent()}</table>;
    }

    return <table className='consoleTable'>{rawConsoleOutput()}</table>;
  };

  const formattedConsoleRow = (firstCell: string, secondCell: string) => (
    <tr>
      <td>{firstCell}</td>
      <td>{secondCell}</td>
    </tr>
  );

  return <div className='outputWrapper'>{formattedConsoleOutput()}</div>;
};

export default Console;
