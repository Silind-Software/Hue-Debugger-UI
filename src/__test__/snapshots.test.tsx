import React from 'react';
import renderer from 'react-test-renderer';
import LoaderSpinner from './../components/LoaderSpinner/LoaderSpinner';
import Console from './../components/Console/Console';
import Authenticate from './../pages/Authenticate/Authenticate';

describe('Components should render correctly', () => {
  const createTree = (fromComponent: React.ReactElement) => renderer.create(fromComponent).toJSON();
  const stubFunc = () => {};

  it('should render LoaderSpinner correctly', () => {
    const tree = createTree(
      <LoaderSpinner isLoading={true} failMessage='Test fail message' backAction={stubFunc} />,
    );

    expect(tree).toMatchSnapshot();
  });

  it('should render Console correctly', () => {
    const tree = createTree(<Console show={true} consoleOutput={[]} toggleConsole={stubFunc} />);

    expect(tree).toMatchSnapshot();
  });

  it('should render Authenticate correctly', () => {
    const tree = createTree(
      <Authenticate setAuthentication={stubFunc} showSweetAlertWarning={stubFunc} />,
    );

    expect(tree).toMatchSnapshot();
  });
});
