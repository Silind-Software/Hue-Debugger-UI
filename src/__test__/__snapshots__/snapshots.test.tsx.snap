// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components should render correctly should render Authenticate correctly 1`] = `
<div
  className="authContainer"
>
  <div
    className="authHeadline"
  >
    Enter bridge IP
  </div>
  <input
    className="authInput"
    onChange={[Function]}
    placeholder="e.g. 192.168.0.100"
    type="text"
    value=""
  />
  <div
    className="authHeadline"
  >
    Developer name
  </div>
  <input
    className="authInput"
    onChange={[Function]}
    placeholder="Choose a developer name"
    type="text"
    value=""
  />
  <div
    className="authButton"
    onClick={[Function]}
  >
    Begin
  </div>
</div>
`;

exports[`Components should render correctly should render Console correctly 1`] = `
<div
  className="consoleWrapper"
  style={
    Object {
      "boxSizing": "border-box",
      "flexShrink": 0,
      "height": "300px",
      "maxHeight": 760,
      "maxWidth": undefined,
      "minHeight": 30,
      "minWidth": undefined,
      "position": "relative",
      "userSelect": "auto",
      "width": "100%",
    }
  }
>
  <div>
    <div
      className="consoleHeader consoleHeaderShow"
      onClick={[Function]}
    >
      Console
    </div>
  </div>
  <div
    className="menuLine"
  >
    <div
      className="menuButton"
      data-title="No formatting"
      onClick={[Function]}
    >
      <i
        className="material-icons"
      >
        notes
      </i>
    </div>
  </div>
  <div
    className="consoleContent"
  />
  <span>
    <div
      className=""
      onMouseDown={[Function]}
      onTouchStart={[Function]}
      style={
        Object {
          "cursor": "row-resize",
          "height": "10px",
          "left": "0px",
          "position": "absolute",
          "top": "-5px",
          "userSelect": "none",
          "width": "100%",
        }
      }
    />
  </span>
</div>
`;

exports[`Components should render correctly should render LoaderSpinner correctly 1`] = `
<div
  className="loaderSpinnerWrapper"
>
  <div
    className="lds-ripple"
  >
    <div />
    <div />
  </div>
  <div
    className="loaderSpinnerText"
  >
    Loading Hue resources
  </div>
</div>
`;
